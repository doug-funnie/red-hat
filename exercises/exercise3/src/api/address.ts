export interface Geo {
  lat: string;
  lng: string;
}

export interface Address {
  string: string;
  street: string;
  suite: string;
  city: string;
  zipcode: string;
  geo: Geo;
}
