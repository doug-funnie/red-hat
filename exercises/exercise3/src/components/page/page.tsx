import glamorous, { Main } from 'glamorous';
import * as React from 'react';
import { Error } from '../error';
import { Header, height as headerHeight } from '../header';

export interface Props {
  title?: string;
  children?: React.ReactNode;
  error?: any;
  onBackButtonPress?(): void;
  onRetry?(): void;
}

const Wrapper = glamorous.div({
  margin: '0 auto',
  maxWidth: 800,
  paddingBottom: 25,
  paddingTop: 15,
  width: '100%',
});

export const Page: React.SFC<Props> = ({
  children,
  title,
  onBackButtonPress,
  error,
  onRetry,
}) => (
  <>
    <Header onBackButtonPress={onBackButtonPress} title={error ? 'Error' : title} />
    <Main marginTop={headerHeight}>
      <Wrapper>
        {!!error ? <Error onRetry={onRetry} /> : children}
      </Wrapper>
    </Main>
  </>
);
