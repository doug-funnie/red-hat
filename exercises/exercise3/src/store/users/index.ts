export * from './users.actions';
export * from './users.reducers';
export * from './users.selectors';
