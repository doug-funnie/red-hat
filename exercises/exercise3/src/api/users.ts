import { Address } from './address';
import { Company } from './company';
import { request } from './request';

const usersBaseUrl = 'http://jsonplaceholder.typicode.com/users';
const avatarBaseUrl = 'https://api.adorable.io/avatars';

export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
}

export function getUsers() {
  return request<User[]>({ url: usersBaseUrl });
}

export function getUser(userId: User['id']) {
  return request<User>({ url: `${usersBaseUrl}/${userId}` });
}

export function getUserAvatarUrl(email: User['email'], size = 35) {
  return `${avatarBaseUrl}/${size}/${email}.png`;
}
