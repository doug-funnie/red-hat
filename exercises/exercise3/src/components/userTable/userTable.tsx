import { User } from 'api/users';
import { Avatar } from 'components/avatar';
import { CellRenderer, Column, EntryClickHandler, KeyExtractor, LabelRenderer, Table } from 'components/table';
import { Div } from 'glamorous';
import * as React from 'react';

type TypedColumn = Column<User>;
type TypedLabelRenderer = LabelRenderer<User>;
type TypedCellRenderer = CellRenderer<User>;
type TypedKeyExtrator = KeyExtractor<User>;
type TypedEntryClickHandler = EntryClickHandler<User>;

interface Props {
  users: User[];
  onUserClick(user: User): void;
}

const nameLabelRenderer: TypedLabelRenderer = () => 'Name';
const nameCellRenderer: TypedCellRenderer =  ({ data }) => (
  <Div display='flex' alignItems='center'>
    <Avatar email={data.email} size={30} />
    <div style={{ marginLeft: 10 }}>
      {data.name}
    </div>
  </Div>
);

const usernameLabelRenderer: TypedLabelRenderer = () => 'Username';
const usernameCellRenderer: TypedCellRenderer = ({ data }) => data.username;

const emailLabelRenderer: TypedLabelRenderer = () => 'Email';
const emailCellRenderer: TypedCellRenderer = ({ data }) => data.email;

const addressLabelRenderer: TypedLabelRenderer = () => 'Address';
const addressCellRenderer: TypedCellRenderer = ({ data }) => (
  <>
    <div>{data.address.street}, {data.address.suite}</div>
    <div>{data.address.city}, {data.address.zipcode}</div>
  </>
);
const columns: TypedColumn[] = [
  { id: 'Name', renderCell: nameCellRenderer, renderLabel: nameLabelRenderer, grow: 2 },
  { id: 'Username', renderCell: usernameCellRenderer, renderLabel: usernameLabelRenderer },
  { id: 'Email', renderCell: emailCellRenderer, renderLabel: emailLabelRenderer, hideOnMobile: true },
  { id: 'Address', renderCell: addressCellRenderer, renderLabel: addressLabelRenderer, grow: 2, hideOnMobile: true },
];
const keyExtractor: TypedKeyExtrator = ({ data }) => data.id;

export class UserTable extends React.Component<Props> {
  private handleEntryClick: TypedEntryClickHandler = ({ data }) => {
    this.props.onUserClick(data);
  }

  public render() {
    return (
      <Table
        data={this.props.users}
        keyExtractor={keyExtractor}
        onEntryClick={this.handleEntryClick}
        columns={columns}
      />
    );
  }
}
