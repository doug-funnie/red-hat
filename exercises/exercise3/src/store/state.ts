import { MapDispatchToProps, MapStateToProps } from 'react-redux';
import { UserListState, UsersState } from './users';

export interface RootState {
  users: UsersState;
  userList: UserListState;
}

export const enum SortDirection {
  asc,
  desc,
}

export const enum FetchStatus {
  none = 'none',
  inProgress = 'inProgress',
  complete = 'complete',
}

const createStateKey = <K extends keyof RootState>(key: K): K => key;
export const usersStateKey = createStateKey('users');
export const userListStateKey = createStateKey('userList');

export type TypedMapStateToProps<InnerProps, OuterProps> = MapStateToProps<InnerProps, OuterProps, RootState>;
export type TypedMapDispatchToProps<DispatchProps> = MapDispatchToProps<DispatchProps, any>;
