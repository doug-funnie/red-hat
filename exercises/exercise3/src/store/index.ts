import { applyMiddleware, createStore } from 'redux';
import { combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { createRequestMiddleware } from './requestMiddleware';
import { RootState, userListStateKey, usersStateKey } from './state';
import { userListReducer, usersReducer } from './users';

export const rootReducer = combineReducers<RootState>({
  [userListStateKey]: userListReducer,
  [usersStateKey]: usersReducer,
});

export function create() {
  return createStore(
    rootReducer,
    applyMiddleware(
      thunk,
      createRequestMiddleware(),
    ),
  );
}
