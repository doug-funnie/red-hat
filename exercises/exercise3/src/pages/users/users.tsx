import { User } from 'api/users';
import { Page } from 'components/page';
import { UserTable } from 'components/userTable';
import * as React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { getUserProfileRoute } from 'routeUtil';
import { FetchStatus, TypedMapDispatchToProps, TypedMapStateToProps } from 'store/state';
import { fetchUsers, selectUserListFetchStatus, selectUserListSorted } from 'store/users';

type OuterProps = RouteComponentProps<{ id: string }>;

interface InnerProps {
  users: User[];
  fetchStatus: FetchStatus;
}

interface DispatchProps {
  fetchUsers(): any;
}

export type Props = InnerProps & DispatchProps & OuterProps;

export const mapStateToProps: TypedMapStateToProps<InnerProps, OuterProps> = (state) => ({
  users: selectUserListSorted(state),
  fetchStatus: selectUserListFetchStatus(state),
});

export const mapDispatchToProps: TypedMapDispatchToProps<DispatchProps> = (dispatch) => bindActionCreators(
  { fetchUsers },
  dispatch,
);

export class Users extends React.Component<Props> {
  public componentDidMount() {
    if (this.props.fetchStatus === FetchStatus.none) {
      this.props.fetchUsers();
    }
  }

  private handleUserClick = (user) => {
    this.props.history.push(getUserProfileRoute(user));
  }

  public render() {
    return (
      <Page title='Users'>
        <UserTable
          users={this.props.users}
          onUserClick={this.handleUserClick}
        />
      </Page>
    );
  }
}

const ConnectedUsers = connect(mapStateToProps, mapDispatchToProps)(Users);

export default ConnectedUsers;
