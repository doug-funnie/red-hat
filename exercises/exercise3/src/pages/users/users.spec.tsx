import { User } from 'api/users';
import { Page } from 'components/page/Page';
import { UserTable } from 'components/userTable/userTable';
import { shallow } from 'enzyme';
import * as React from 'react';
import { FetchStatus } from 'store/state';
import '../../../../../testUtils/addJestMatchers';
import { Props, Users } from './users';

const props: Props = {
  users: [{
    id: 1,
  } as any],
  fetchStatus: FetchStatus.none,
  fetchUsers: jest.fn(),
  match: {} as any,
  location: {} as any,
  history: {
    push: jest.fn(),
  } as any,
};

it('only calls fetchUsers if fetchStatus is none', () => {
  shallow(<Users {...props} fetchStatus={FetchStatus.complete} />);
  shallow(<Users {...props} fetchStatus={FetchStatus.inProgress} />);
  expect(props.fetchUsers).not.toBeCalled();
  shallow(<Users {...props} />);
  expect(props.fetchUsers).toBeCalled();
});

it('renders title as Users', () => {
  const view = shallow(<Users  {...props} />);
  expect(view.find(Page).props().title).toBe('Users');
});

it('calls history.push onUserClick', () => {
  const view = shallow(<Users  {...props} />);
  view.find(UserTable).props().onUserClick({ id: 1 } as User);
  expect(props.history.push).toBeCalledWithSnapshot();
});
