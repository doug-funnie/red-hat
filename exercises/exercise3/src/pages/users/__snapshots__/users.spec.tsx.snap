// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls history.push onUserClick 1`] = `
Array [
  Array [
    "/users/1",
  ],
]
`;
